<html>
  <head>
    <link rel = "stylesheet" href = "default.css">
	<title>Transposable Element Finder (TEF)</title>
  </head>
  <body>
    <h1>Transposable Element Finder (TEF)<font size = "4"> - Detection of active transposable elements from NGS data</font></h1>
    TEF is a high-performance detector of active transposable elements from short-reads of next generation sequencer, based on detection of Target Site Duplication (TSD) at the transposition.
    <h2>News</h2>
	<ul>
    <li><a href = "https://www.nature.com/articles/s41467-023-41023-0" target = "_blank">Result of TEF is cited by a Nature communications paper</a>
	<li><a href = "https://zenodo.org/records/11069379" target = "_blank">List of head and tail seqeucne pairs of transposable element in rice </a><font color = "red"><b>Update (2024-04-26)</b></font>
	<li><a href = "arabidopsis.html" target = "_blank">Detection of transposed TEs in <i>Arabidopsis thaliana</i> herbarium specimen</a> <font color = "red"><b>Update (2023-12-15)</b></font>
    </ul>
    <h2>Tutrial</h2>
	<ul>
    <li>Scripts and Readme: <a href = "https://github.com/akiomiyao/tef/" target = "_blank">https://github.com/akiomiyao/tef/</a><br><br>
	<li>Download the souce code from 'Code' pulldown menu or clone the code in your local machine.<br>
	<pre><code>git clone https://github.com/akiomiyao/tef.git</code></pre>
	'tef' directory will be cleated. All codes will be saved into the tef directory.
	<li>Download fastq files<br>
The following explanation uses nucleotide sequence data obtained from two individuals of <i>Drosophila melanogaster</i>, which are relatively small in size and require short analysis time.<br>
Download from <a href = "https://ddbj.nig.ac.jp/search" target = "_blank">DDBJ DRA page</a>,<br>
<a href = "https://ddbj.nig.ac.jp/resource/sra-run/SRR823377" target = "_blank">https://ddbj.nig.ac.jp/resource/sra-run/SRR823377</a><br>
&nbsp;&nbsp;<a href = "https://ddbj.nig.ac.jp/public/ddbj_database/dra/fastq/SRA071/SRA071525/SRX262892/SRR823377_1.fastq.bz2">SRR823377_1.fastq.bz2</a> and
<a href = "https://ddbj.nig.ac.jp/public/ddbj_database/dra/fastq/SRA071/SRA071525/SRX262892/SRR823377_2.fastq.bz2">SRR823377_2.fastq.bz2</a><br>
<a href = "https://ddbj.nig.ac.jp/resource/sra-run/SRR823382" target = "_blank">https://ddbj.nig.ac.jp/resource/sra-run/SRR823382</a><br>
&nbsp;&nbsp;<a href = "https://ddbj.nig.ac.jp/public/ddbj_database/dra/fastq/SRA071/SRA071525/SRX262914/SRR823382_1.fastq.bz2">SRR823382_1.fastq.bz2</a> and
<a href = "https://ddbj.nig.ac.jp/public/ddbj_database/dra/fastq/SRA071/SRA071525/SRX262914/SRR823382_2.fastq.bz2">SRR823382_2.fastq.bz2</a><br>
Files with bz2 format can also be analyzed by tef.pl. <br><br>
Or download fastq.gz files from NCBI using fastq-dump.<br>
'fastq-dump' can be installed by apt command (Ubuntu).<br>
<i>e.g.</i> apt install fastq-dump<br>
fastq-dump --split-files SRR823377 SRR823382<br>
	<li>First run of tef.pl
	<pre><code>
git clone https://github.com/akiomiyao/tef.git
cd tef
mkdir SRR823377 SRR823382
mkdir SRR823377/read
mkdir SRR823382/read
mv Download/SRR823377.fastq.gz SRR823377/read
mv Download/SRR823382.fastq.gz SRR823382/read
mkdir dmel
cd dmel
curl -O https://flybase.org/genomes/Drosophila_melanogaster/dmel_r6.64_FB2025_03/fasta/dmel-all-chromosome-r6.64.fasta.gz
cd ..
perl tef.pl a=SRR823377,b=SRR823382,ref=dmel
</code></pre>
	Options are separated by comma. Do not add space after comma.
	<li>At the first run of tef.pl. 'config' file is created in the reference directory.<br>
Edit config file in tef/dmel directory. Scaffold may be not required.<br>
If names of scaffold are changed to 'NOP', nucleotide seqeunce of scaffold will be ignored.<br>
For example,<pre>
2L
2R
3L
3R
4
X
Y
NOP
NOP
NOP
NOP
NOP
NOP
continue to line number 1870.
</pre>
	<li>If you feel hard to change to NOP, open config file by Excel, and paste 'NOP' in all scaffold cells, and save 'config' in the 'tef/dmel' directory as a text file.<br>
	<a href = "config" target = "_blank">Link to 'config' file for <i>Drosophila melanogaster</i>.<a>
	<li>Run again tef.pl
<pre><code>perl tef.pl a=SRR823377,b=SRR823382,ref=dmel</code></pre>
	Default is mode of 'Junction method'.
	<li>For the analysis 'TSD method', specify the value of 'tsd_size'.<br>
For example,
<pre><code>perl tef.pl a=SRR823377,b=SRR823382,ref=dmel,tsd_size=8</code></pre>
	<li>Results are saved in the 'SRR823377' (specified with 'a') directory.
	<pre>
child/
count.20/
count.8/
junction_method.genotype.SRR823377.SRR823382
junction_method.log
junction_method.summary.SRR823377.SRR823382
junction_method.tepos.SRR823377.SRR823382
junction_method.verify.SRR823377.SRR823382
read/
split/
tmp/
tsd_method.genotype.SRR823377.SRR823382.8
tsd_method.log.8
tsd_method.pair.SRR823377.SRR823382.8
tsd_method.summary.SRR823377.SRR823382.8
tsd_method.tepos.SRR823377.SRR823382.8
tsd_method.verify.SRR823377.SRR823382.8
	</pre>
	<li>Following is a part of genotype file of TSD method.<pre>
                                                                                                                                                                   SRR823377              SRR823382             
Accession   Chr.  Head Junction  Tail Junction  Direction TE Head                 TE Tail                 Head Flanking          TSD        Tail Flanking          Head   Tail  No-ins.   Head    Tail   No-Ins.  Genotype
SRR823377   3L    11009306       11009313       r         CATGATGAAATAACATAAGG    CCTTATGTTATTTCATCATG    CGGCGAATGAGAGGCCAACA   GGCCAACA   GGCCAACAGTCGGCACTAAC   5      7     15        0       0      26       H
SRR823377   3L    11009313       11009306       f         CATGATGAAATAACATAAGG    CCTTATGTTATTTCATCATG    GTTAGTGCCGACTGTTGGCC   TGTTGGCC   TGTTGGCCTCTCATTCGCCG   7      5     17        0       0      23       H
SRR823377   3L    13569996       13570003       r         CAGAGAACTGCAAGGGTGGC    TGGCGGGCTGCAGTTCTCTG    ATGGTGGTGTGCATGTGGGG   ATGTGGGG   ATGTGGGGTAACCTGCCCAA   25     15    0         0       0      16       M
SRR823377   3L    13570003       13569996       f         CAGAGAACTGCAGCCCGCCA    GCCACCCTTGCAGTTCTCTG    TTGGGCAGGTTACCCCACAT   CCCCACAT   CCCCACATGCACACCACCAT   15     25    0         0       0      13       M
SRR823377   3L    16020833       16020840       r         CAGAGAACTGCAAGGGTGGC    TGGCGGGCTGCAGTTCTCTG    CAACAGATGGGCTCCCGGAT   TCCCGGAT   TCCCGGATCGCCGGTCGCCG   19     10    0         0       0      27       M
SRR823377   3L    16020840       16020833       f         CAGAGAACTGCAGCCCGCCA    GCCACCCTTGCAGTTCTCTG    CGGCGACCGGCGATCCGGGA   ATCCGGGA   ATCCGGGAGCCCATCTGTTG   10     19    0         0       0      27       M
SRR823377   3L    23672618       23672625       r         CATGATGAAATAACATAAGG    CCTTATGTTATTTCATCATG    CCGCTGCGCTTGGGATAGCC   GGATAGCC   GGATAGCCTAACTAATTAGA   9      7     395       0       0      315      H
SRR823377   3L    23672619       23672626       r         CCATGATGAAATAACATAAG    ACCTTATGTTATTTCATCAT    TCCGCTGCGCTTGGGATAGC   GGGATAGC   GGGATAGCCTAACTAATTAG   9      7     397       0       0      325      H
SRR823377   3L    23672625       23672618       f         CATGATGAAATAACATAAGG    CCTTATGTTATTTCATCATG    TCTAATTAGTTAGGCTATCC   GGCTATCC   GGCTATCCCAAGCGCAGCGG   7      9     64        0       0      71       H
SRR823377   3L    23672626       23672619       f         ATGATGAAATAACATAAGGT    CTTATGTTATTTCATCATGG    CTAATTAGTTAGGCTATCCC   GCTATCCC   GCTATCCCAAGCGCAGCGGA   7      9     63        0       0      71       H
</pre>
The result file is tab separated text file, and can be opened by Excel.<br>
The CATGATGAAATAACATAAGG-CCTTATGTTATTTCATCATG head-tail pair is <i>P-</i>element.<br>
Numbers of Head, Tail, and No-ins. are read counts containing junction of head, junction of tail, and No-insertion (wild-type) from SRR823377 and SRR823382.<br>
Forward and complementary seqeunces of TE Head and TE Tail are listed.<br>
If counts of Head, Tail and No-ins. are detected in SRR823377 and only counts of No-Ins. are detected in SRR823382, TE is transposed into only SRR823377 and genotype is Heterozygous (shown in 'H').<br>
If counts of Head, Tail are detected and no count for No-ins. in SRR823377 and only counts of No-Ins. are detected in SRR823382, TE is transposed into only SRR823377 and genotype is Homozygous (shown in 'M').<br>
The high counts of No-insertion indicats that the transposed region is in the repetitive region.<br>
If pair of adjuscent bases of TSD is complementary, tef.pl detects both positions as junctions. <i>e.g.,</i> 3L 23672618 and 23672619.
</ul>

    <h2>Citing TEF</h2>
    Miyao, A., Yamanouchi, U. Transposable element finder (TEF): finding active transposable elements from next generation sequencing data. BMC Bioinformatics 23, 500 (2022). <a href = "https://doi.org/10.1186/s12859-022-05011-3" target = "_blank">https://doi.org/10.1186/s12859-022-05011-3</a><br>
    <h2>Contact</h2>
    Akio Miyao, Ph.D (<a href = "mailto:miyao@affrc.go.jp">miyao@affrc.go.jp</a>)<br>
  </body>
</html>
